### PR TITLE
Headsets

### PR DESCRIPTION
**Pull request by DrSmugleaf**
Supersedes #1598

### CODE REVIEW

**Comment by DrSmugleaf** [comment]
Reviews addressed

File: Resources/Prototypes/Entities/Clothing/Earpieces/ears.yml [hunk 1]
```diff
@@ -1,16 +1,230 @@
...
+  components:
+    - type: Headset
+
+
+## Earpiece Headsets
+- type: entity
+  parent: BaseHeadset
+  id: HeadsetRadioEars
+  name: headset
+  description: The radio to keep up to date in real time with fun onboard station activities
```
Comment (exp111) [inline Resources/Prototypes/Entities/Clothing/Earpieces/ears.yml:29]: I think descriptions are inherited

File: Resources/Prototypes/Entities/Clothing/Earpieces/ears.yml [hunk 2]
```diff
@@ -1,16 +1,230 @@
...
+    state: headset
+  - type: Icon
+    sprite: Objects/Devices/communication.rsi
+    state: headset
+
+- type: entity
+  parent: BaseHeadset
+  id: HeadsetBase
+  name: cargo headset
+  description:
```
Comment (exp111) [inline Resources/Prototypes/Entities/Clothing/Earpieces/ears.yml:42]: Most of the headsets have no description. Is this a future junior job?

**Review by exp111 — COMMENTED** [review]
The mime is missing a headset as starting gear (apparently the service headset?).
Also they currently don't work, but you know that

File: Content.Server/GameObjects/Components/HandheldRadioComponent.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            _chatManager.EntitySay(Owner, message);
+        }
+
+        public bool UseEntity(UseEntityEventArgs eventArgs)
+        {
+            RadioOn = !RadioOn;
+
+            Owner.PopupMessage(eventArgs.User, RadioOn
+                ? "The radio is now on."
+                : "The radio is now off.");
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/GameObjects/Components/HandheldRadioComponent.cs:80]: This should also be localised.

File: Content.Server/Chat/ChatManager.cs [hunk 4]
```diff
@@ -115,15 +120,31 @@ public void EntitySay(IEntity source, string message)
             var pos = source.Transform.Coordinates;
             var clients = _playerManager.GetPlayersInRange(pos, VoiceRange).Select(p => p.ConnectedClient);
 
+            if (message.StartsWith(';'))
+            {
+                message = message.Substring(1);
```
Comment (exp111) [inline Content.Server/Chat/ChatManager.cs:125]: Maybe trim the start of the string? Many people will write "; hello" or ";hello" and we probably don't want the space there.

File: Content.Server/GameObjects/Components/HeadsetComponent.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                if (!container.Owner.TryGetComponent(out IActorComponent actor))
+                    return;
+
+                var playerChannel = actor.playerSession.ConnectedClient;
+
+                var msg = _netManager.CreateNetMessage<MsgChatMessage>();
+
+                msg.Channel = ChatChannel.Radio;
+                msg.Message = message;
+                msg.MessageWrap = Loc.GetString("[{0}] {1} says, {{0}}", channel, source.Name);
```
Comment (exp111) [inline Content.Server/GameObjects/Components/HeadsetComponent.cs:77]: grafik
Add the quotes

File: Content.Server/GameObjects/Components/HeadsetComponent.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                _netManager.ServerSendMessage(msg, playerChannel);
```
Comment (exp111) [inline Content.Server/GameObjects/Components/HeadsetComponent.cs:78]: The player itself doesn't get the message, which would be nice as a confirmation "yes you really did send the msg on the radio"

**Review by exp111 — COMMENTED** [review]
Couldn't test the handheld radio
grafik

```cs
[ERRO] go.comp.sprite: Unable to load RSI '/Textures/Objects/Devices/communication.rsi'. Trace:
SixLabors.ImageSharp.UnknownImageFormatException: Image cannot be loaded. Available decoders:
 - GIF : GifDecoder
 - TGA : TgaDecoder
 - JPEG : JpegDecoder
 - PNG : PngDecoder
 - BMP : BmpDecoder

   at SixLabors.ImageSharp.Image.Load[TPixel](Configuration configuration, Stream stream, IImageFormat& format)
   at SixLabors.ImageSharp.Image.Load[TPixel](Stream stream)
   at Robust.Client.ResourceManagement.RSIResource.Load(IResourceCache cache, ResourcePath path) in D:\Visual Studio\Github\ss14\review-space-station-14\RobustToolbox\Robust.Client\ResourceManagement\ResourceTypes\RSIResource.cs:line 86
   at Robust.Client.ResourceManagement.ResourceCache.GetResource[T](ResourcePath path, Boolean useFallback) in D:\Visual Studio\Github\ss14\review-space-station-14\RobustToolbox\Robust.Client\ResourceManagement\ResourceCache.cs:line 36
   at Robust.Client.GameObjects.SpriteComponent.ExposeData(ObjectSerializer serializer) in D:\Visual Studio\Github\ss14\review-space-station-14\RobustToolbox\Robust.Client\GameObjects\Components\Renderable\SpriteComponent.cs:line 1072
[ERRO] go.comp.sprite: Layer has no RSI to load states from.cannot use 'state' property. Prototype: 'handrad'
[ERRO] root: Exception while loading resource Robust.Client.ResourceManagement.RSIResource at '/Textures/Objects/Devices/communication.rsi', no fallback available
```

**Comment by DrSmugleaf** [comment]
Reviews addressed.

**Review by ShadowCommander — CHANGES_REQUESTED** [review]
It'd be nice for IActivate to be implemented on the HandheldRadioComponent, so that people don't have to pick it up to turn it on/off.

Later on multiple channels need to be implemented so that `:e` and `:q` work for specifying which preset frequency.
GUI for changing the frequency. (Usually on the handheld radio. Sometimes on the headset depending on codebase.)
Examine tooltip to see available radio prefixes.

File: Content.Server/Interfaces/IListen.cs [hunk 7]
```diff
@@ -1,14 +1,20 @@
...
-    public interface IListen
+    public interface IListen : IComponent
     {
-        void HeardSpeech(string speech, IEntity source);
+        int ListenRange { get; }
 
-        int GetListenRange();
+        bool CanHear(string message, IEntity source);
+
+        void Broadcast(string message, IEntity speaker);
```
[the remark this answers is not in the document]
Comment (DrSmugleaf) [reply Content.Server/Interfaces/IListen.cs]: Done, they are both called CanListen and Listen now.

**Comment by DrSmugleaf** [comment]
> It'd be nice for IActivate to be implemented on the HandheldRadioComponent, so that people don't have to pick it up to turn it on/off.
> 
> Later on multiple channels need to be implemented so that `:e` and `:q` work for specifying which preset frequency.
> GUI for changing the frequency. (Usually on the handheld radio. Sometimes on the headset depending on codebase.)
> Examine tooltip to see available radio prefixes.

All done except for the multiple prefixes and GUI. Should I do those in this pr as well?

**Review by ShadowCommander — APPROVED** [review]

**Comment by ShadowCommander** [comment]
Should be fine like this. Channels, GUI, and examine can be added in a separate PR.

**Pull request merged**